sn_transfers/src/cashnotes/output_purpose.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2024 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use bls::{Ciphertext, PublicKey, SecretKey};
+use serde::{Deserialize, Serialize};
+use xor_name::XorName;
+
+use crate::{DerivationIndex, Hash, Result, TransferError};
+
+const CUSTOM_OUTPUT_PURPOSE_SIZE: usize = 64;
+
+/// The attached metadata or reason for which a Spend was spent
+#[derive(Default, Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub enum OutputPurpose {
+    #[default]
+    None,
+    /// Reference to network data
+    NetworkData(XorName),
+    /// Custom field for any application data
+    Custom(#[serde(with = "serde_bytes")] [u8; CUSTOM_OUTPUT_PURPOSE_SIZE]),
+    /// For Network Royalty Fee
+    RoyaltyFee(DerivationIndex),
+
+    /// Beta only feature to track rewards
+    /// Discord username encrypted to the Foundation's pubkey with a random nonce
+    BetaRewardTracking(DiscordNameCipher),
+}
+
+impl OutputPurpose {
+    pub fn hash(&self) -> Hash {
+        match self {
+            Self::None => Hash::default(),
+            Self::NetworkData(xor_name) => Hash::hash(xor_name),
+            Self::Custom(bytes) => Hash::hash(bytes),
+            Self::RoyaltyFee(derivation_index) => Hash::hash(&derivation_index.0),
+            Self::BetaRewardTracking(cypher) => Hash::hash(&cypher.cipher),
+        }
+    }
+
+    pub fn create_reward_tracking_purpose(input_str: &str) -> Result<Self> {
+        let input_pk = crate::PAYMENT_FORWARD_PK.public_key();
+        Ok(Self::BetaRewardTracking(DiscordNameCipher::create(
+            input_str, input_pk,
+        )?))
+    }
+
+    pub fn get_sender_hash(&self, sk: &SecretKey) -> Option<Hash> {
```

Comment:
Rust's convention is not to start getters with `get_`. E.g. `decrypt_sender` or `decrypt_discord_cypher`.